src/highdicom/io.py

```diff
@@ -232,15 +233,18 @@ class ImageFileReader(object):
 
     """
 
-    def __init__(self, filename: Union[str, Path, DicomFileLike]):
+    def __init__(self, filename: Union[str, Path, DicomFileLike, DicomBytesIO]):
         """
         Parameters
         ----------
         filename: Union[str, pathlib.Path, pydicom.filebase.DicomfileLike]
             DICOM Part10 file containing a dataset of an image SOP Instance
 
         """
-        if isinstance(filename, DicomFileLike):
+        if isinstance(filename, DicomBytesIO):
+            self._fp = filename
+            self._filename = None
+        elif isinstance(filename, DicomFileLike):
```

Comment:
I don't get this edit. Do you want to remove all the type checks on `filename` ?

Reply:
I'm also confused by this...

There are three cases that we need to consider:
- The plain str/`pathlib.Path` case
- The `DicomBytesIO` case (addressed in this PR)
- The `DicomFileLike` case where the `DicomFileLike` object is wrapped around an open file on disk. This case is currently broken on master and not fixed by this branch.

I was able to get all three cases to work with the following as my full `__init__` method:

```python
    def __init__(self, filename: Union[str, Path, DicomFileLike, DicomBytesIO]):
        """
        Parameters
        ----------
        filename: Union[str, pathlib.Path, pydicom.filebase.DicomfileLike]
            DICOM Part10 file containing a dataset of an image SOP Instance

        """
        if isinstance(filename, DicomFileLike):
            fp = filename
            if isinstance(filename, DicomBytesIO):
                self._fp = filename
                self._filename = None
            else:
                self._fp = fp
                self._filename = Path(fp.name)
        elif isinstance(filename, (str, Path)):
            self._filename = Path(filename)
            self._fp = None
        else:
            raise TypeError(
                'Argument "filename" must either an open DICOM file object or '
                'the path to a DICOM file stored on disk.'
            )
        self._metadata = None
```

However to get this to work I had to get rid of the tests on `is_little_endian` and `is_implicit_vr`. I don't think these are necessary (why would these characteristics not be read correctly from the file) but in the spirit of Chesterton's fence I'd rather understand from @hackermd why they were there in the first place before ripping them out.

Reply:
My point is that `DicomBytesIO` is a subtype of `DicomFileLike` and we only need to differentiate between the two types for setting `self._filename`. The other checks and setting of `self._fp` can be performed for both. I intended the checks to stay in place.

Reply:
Understood, but as far as I can tell these checks never worked in the first place (and we missed this because we didn't have a test case for them). The checks did this:

1. Get values of `is_little_endian` and `is_implicit_vr` by reading the file meta using the `_check_file_format` method (which in turn uses the `pydicom.filereader.read_partial` method).
2. Get the same values by directly accessing (Python) attributes of the same name directly on the `fp` object, which will be a `DicomFileLike` object.
 
I'm currently failing to understand two points:
- Why would (2.) above ever work, since the `DicomFileLike` class does not have `is_little_endian` and `is_implicit_vr` attributes? I have played around with this and it always fails with an `AttributeError`
- If we imagine (2.) above did somehow work, under what circumstances would the results be different from the values obtained by (1.) above? In other words, why is check necessary?

It seems I am missing something big so I am unable to advise further except for just removing these checks, which as far as I can tell never actually worked and are unnecessary.

Reply:
Making some progress on understanding the confusion. What I said above is incorrect, `DicomFileLike` objects *do* have an `is_implicit_vr` attribute. Further, it does have an [`is_little_endian` property](https://github.com/pydicom/pydicom/blob/f7fdf0f12b235a315f850d58ab4e335a35f336b5/pydicom/filebase.py#L140), but this fails to return a value unless the [`is_little_endian` setter](https://github.com/pydicom/pydicom/blob/f7fdf0f12b235a315f850d58ab4e335a35f336b5/pydicom/filebase.py#L144) has previously explicitly been called on the object. This may actually be a bug in pydicom. This is what is causing my tests to fail whenever I tried it.

To demonstrate with the various ways of constructing a `DicomFileLike`

```python
from pydicom.filebase import DicomFile, DicomFileLike, DicomBytesIO
from pydicom.data import get_testdata_file

ct_file = get_testdata_file("ct_small.dcm")
print(ct_file)

# Using DicomFile function
dcm = DicomFile(ct_file, mode='rb')
assert isinstance(dcm, DicomFileLike)
print(dcm.is_implicit_VR)  # works
print(dcm.is_little_endian)  # fails

# Using DicomFileLike constructor on an open file
dcm = DicomFileLike(open(ct_file, mode='rb'))
assert isinstance(dcm, DicomFileLike)
print(dcm.is_implicit_VR) # works
# print(dcm.is_little_endian) # fails

# Using DicomBytesIO
dcm = DicomBytesIO(open(ct_file, "rb").read())
assert isinstance(dcm, DicomFileLike)
print(dcm.is_implicit_VR) # works
# print(dcm.is_little_endian) # fails
```

So in order to make this work as expected, the user would have to explicitly set the `is_little_endian` attribute before passing the object to the `ImageFileReader` class. Is this your intent @hackermd ? I would guess not...

Reply:
Thanks for taking the deep-dive @CPBridge. Not sure whether it is a pydicom bug or feature, but I agree with you that the implementation of the `is_little_endian` logic is problematic. I would be fine with omitting the checks.